qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py

```diff
@@ -413,20 +410,33 @@ def add_new_detector_mask(self, new_mask: list[bool]) -> str:
         return new_key
 
     def apply_detector_masks(self, mask_keys: list[str]) -> None:
-        if not mask_keys:
-            mask_total = self._is_masked_in_ws
-        else:
-            mask_total = np.logical_or.reduce([self._cached_masks_map[key] for key in mask_keys])
-        self._is_masked = mask_total
+        ws_masks = [ws.extractY().flatten() for ws in self.get_mask_workspaces_in_ads() if ws.name() in mask_keys]
+        cached_masks = [self._cached_masks_map[key] for key in mask_keys if key in self._cached_masks_map.keys()]
+
+        if not ws_masks and not cached_masks:
+            self._is_masked = self._is_masked_in_ws
+            self._detector_is_picked[~self.is_pickable] = False
+            return
+
+        total_mask = np.logical_or.reduce(ws_masks + cached_masks)
+        self._is_masked = total_mask
         self._detector_is_picked[~self.is_pickable] = False
 
+    def clear_stored_masks(self) -> None:
+        self._cached_masks_map.clear()
+
+    @property
+    def cached_masks_keys(self) -> list[str]:
+        return list(self._cached_masks_map.keys())
+
     def save_mask_workspace_to_ads(self) -> None:
-        name_exported_ws = f"instrument_view_mask_{self._workspace.name()}"
         for i, v in enumerate(self._is_masked):
             self._mask_ws.dataY(i)[:] = v
 
         xmin, xmax = self._integration_limits
-        ExtractMaskToTable(self._mask_ws, Xmin=xmin, Xmax=xmax, OutputWorkspace=name_exported_ws)
+        # TODO: Figure out naming convention
+        ExtractMaskToTable(self._mask_ws, Xmin=xmin, Xmax=xmax, OutputWorkspace="MaskTable")
+        CloneWorkspace(self._mask_ws, OutputWorkspace="MaskWorkspace")
```

Comment:
Is this `TODO` resolved?

Reply:
No problem, you can get to that later. Just checking it wasn't a temporary thing you forgot to remove